File: 3-LongestSubStrWithoutRepeatingChar/longest_sub_str.cpp
#include <string>
#include <algorithm>

class Solution
{
public:
    int lengthOfLongestSubstring(std::string s)
    {
        if (s.empty())
            return 0;

        int charIndex[128];
        std::fill_n(charIndex, 128, -1);
        int maxLength = 0;
        int startIndex = 0;

        for (int endIndex = 0; endIndex < s.size(); ++endIndex)
        {
            int charCode = static_cast<int>(s[endIndex]);
            startIndex = std::max(startIndex, charIndex[charCode] + 1);
            charIndex[charCode] = endIndex;
            maxLength = std::max(maxLength, endIndex - startIndex + 1);
        }

        return maxLength;
    }
};

File: 3-LongestSubStrWithoutRepeatingChar/longest_sub_str.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0

        char_index = [-1] * 128

        max_length = 0
        start_index = 0

        for end_index, char in enumerate(s):
            char_code = ord(char)
            start_index = max(start_index, char_index[char_code] + 1)
            char_index[char_code] = end_index
            max_length = max(max_length, end_index - start_index + 1)

        return max_length
